marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/main.sjs

```diff
@@ -79,7 +80,7 @@ function main(contentSequence, options, stepExecutionContext) {
       try {
         arrayOfInstanceArrays = xqueryLib.dataHubMapToCanonical(instance, mappingURIforXML, mappingParams, {"format":outputFormat});
       } catch (e) {
-        throw Error(e);
+        throw Error(mappingLibrary.parseMappingError(e));
```

Comment:
This is confusing - parseMappingError will either return a string or an error object. Shouldn't this always return a string so it can be passed to the Error constructor?

Reply:
@rjrudin 
We were passing the error object as argument to 'Error' constructor (till 5.5). I was just following the same pattern except in cases where the error is one that's specified in the jira stories in which case I construct a more meaningful message and pass that as argument to 'Error' constructor. Should I do something differently ? If the error occurring is not listed in one of the stories, what string should be passed to 'Error' constructor ?

Reply:
I added a comment above that "throw Error(e)" part to explain it:

```
// For not-yet-known reasons, catching this error and then simply rethrowing it causes the MappingTest JUnit class
      // to pass due to the "message" part of the JSON in the stepOutput containing the expected output when this is done.
```

This seems pretty brittle - and perhaps the MappingTest shouldn't really be caring about fairly technical stuff within the error contents. It certainly seems simpler if we just do "throw Error(message)". 

Can you take a look at the MappingTest test that makes an assertion on the contents of "stepOutput" and see if that can be reasonably adjusted based on the output of "throw Error(message)"?

Reply:
So should I just 

1. 'throw Error(message)' in case where 'e' is one of the errors listed in jira stories  and
2. 'throw e' in case of other errors (the ones not listed in jira stories)

(and fix MappingTest) ?

Reply:
I like that - "throw Error(message)", else "throw e".

Reply:
I am seeing that
1. When error object 'e' is passed as argument to 'Error' constructor (current behavior), stepOutput.message is 

"RESTAPI-SRVEXERR: /*:stylesheet/*:template[2] <xsl:element name=\"instance:mapping0Instances\"> <xsl:for-each select=\"/\"> <xsl:element name=\"instance:entityInstance\"> <xsl:element name=\"uri\"> <xsl:value-of select=\"$URI\"/></xsl:element><xsl:element name=\"value\"> <xsl:call-template name=\"mapping0-Order\"/></xsl:element></xsl:element></xsl:for-each></xsl:element> -- Extension Error:  code: 400 message: The pattern 'YYYY.MM.DD' cannot be applied to the value '01/08/1996' document: %3"
<img width="756" alt="Screen Shot 2021-06-14 at 3 16 24 PM" src="https://user-images.githubusercontent.com/5034960/121966627-cf46b880-cd23-11eb-8d68-05e00af244d5.png">


2. When we just 'throw e', stepOutput.message is just "Extension Error"
<img width="752" alt="Screen Shot 2021-06-14 at 3 17 07 PM" src="https://user-images.githubusercontent.com/5034960/121966653-dc63a780-cd23-11eb-8b6b-7fcdc2216fe0.png">


Doing the latter (throw e) also affects the message displayed in HC when step fails as can be seen in the screenshots in each cases

Reply:
It seems this weirdness of needing "throw Error(e)" is specific to this block of code here. Thus, I think the mapping-lib function should ignore this and have a scope of "I'll try to extract a message from this, and if I cannot, I'll return null so the client can decide what to do". That way, the mapping-lib function is only returning a string or null, and not a string or an error object. 

The client code would then be (pseudocode):

```
const message = mappingLib.getHumanFriendlyErrorMessage(error);
if (message) throw Error(message) else throw Error(error);
```

Reply:
And actually, i think "extract" is a better verb than "get", as "extract" implies that we're going to dig around in the error and see if we can come up with a nice error message. And "human" is redundant here - "extractFriendlyErrorMessage" should do the trick.